server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class represents a node's view of the history of which nodes have been elected master over the last 30 minutes. It is kept in
+ * memory, so when a node comes up it does not have any knowledge of previous master history before that point.
+ */
+public class MasterHistory implements ClusterStateListener, Writeable, Writeable.Reader<MasterHistory> {
+    private List<TimeAndMaster> masterHistory;
+    Supplier<Long> nowSupplier = System::currentTimeMillis; // Can be changed for testing
+
+    public MasterHistory(ClusterService clusterService) {
+        this.masterHistory = new ArrayList<>();
+        clusterService.addListener(this);
+    }
+
+    public MasterHistory(StreamInput streamInput) throws IOException {
+        this.masterHistory = streamInput.readList(TimeAndMaster::new);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+        DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+        if (currentMaster == null || currentMaster.equals(previousMaster) == false || masterHistory.isEmpty()) {
+            masterHistory.add(new TimeAndMaster(nowSupplier.get(), currentMaster));
+        }
+        removeOldMasterHistory();
```

Comment:
As this is called from the cluster change notified thread and potentially from a public method (ie. `getMostRecentNonNullMaster`) this poses a multi-thread safety issue.

We either choose a thread-safe collection to store the history (ie. `CopyOnWriteArrayList` ) or we just purge the old history in the `clusterChanged` call (ie. have one writer against the history List).

I believe the latter is a good candidate here. What do you think?

Reply:
Good point! I forgot to come back to that. If we only purge in `clusterChanged` I'll need to change the read methods to ignore anything old, since we could probably go a very long time between a cluster changed event. But that would work.

Reply:
I wound up protecting write access from multiple threads because I thought it would be easier to read the code (vs another check for all reads to ignore anything older than 30 minutes). Let me know what you think.